kimchi/src/circuits/lookup/index.rs

```diff
@@ -217,21 +215,64 @@ impl<F: PrimeField + SquareRootField> LookupConstraintSystem<F> {
                     lookup_info.selector_polynomials_and_tables(domain, gates);
 
-                //~ 3. Concatenate runtime lookup tables with the ones used by gates
-                let mut lookup_tables: Vec<_> = gate_lookup_tables
+                // Checks whether an iterator contains any duplicates, and if yes, raises
+                // a corresponding LookupTableIdCollision error.
+                fn check_id_duplicates<'a, I: Iterator<Item = &'a i32>>(
+                    iter: I,
+                    msg: &str,
+                ) -> Result<(), LookupError> {
+                    use itertools::Itertools;
+                    match iter.duplicates().collect::<Vec<_>>() {
+                        dups if !dups.is_empty() => Err(LookupError::LookupTableIdCollision {
+                            collision_type: format!("{}: {:?}", msg, dups).to_string(),
+                        }),
+                        _ => Ok(()),
+                    }
+                }
+
+                // If there is a gate using a lookup table, this table must not be added
+                // explicitly to the constraint system.
+                let fixed_gate_joint_ids: Vec<i32> = fixed_lookup_tables
+                    .iter()
+                    .map(|lt| lt.id())
+                    .chain(gate_lookup_tables.iter().map(|lt| lt.id()))
+                    .collect();
+                check_id_duplicates(
+                    fixed_gate_joint_ids.iter(),
+                    "duplicates between fixed given and fixed from-gate tables",
+                )?;
+
+                //~ 3. Concatenate explicit runtime lookup tables with the ones (implicitly) used by gates.
+                let mut lookup_tables: Vec<LookupTable<_>> = fixed_lookup_tables
                     .into_iter()
-                    .chain(lookup_tables)
+                    .chain(gate_lookup_tables)
                     .collect();
 
-                let mut has_table_id_0 = false;
+                let fixed_lookup_tables_ids: Vec<i32> =
+                    lookup_tables.iter().map(|lt| lt.id()).collect();
+                check_id_duplicates(
+                    fixed_lookup_tables_ids.iter(),
+                    "fixed lookup table duplicates",
+                )?;
 
                 // if we are using runtime tables
                 let (runtime_table_offset, runtime_selector) =
                     if let Some(runtime_tables) = &runtime_tables {
+                        let runtime_tables_ids: Vec<i32> =
+                            runtime_tables.iter().map(|rt| rt.id).collect();
+                        check_id_duplicates(runtime_tables_ids.iter(), "runtime table duplicates")?;
+                        check_id_duplicates(
+                            runtime_tables_ids
+                                .iter()
+                                .chain(fixed_lookup_tables_ids.iter()),
+                            "duplicates between runtime and fixed tables",
```

Comment:
We explicitly don't want this. This should never have landed without backing this out.